### PR TITLE
REPLAY-1333 SR - add capability to record TabViews

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

In this PR we are adding a new capability to the SR recorder to be able to record `TabViews` by introducing the `TabViewMapper` into the `dd-sdk-android-session-replay-material` module. Because the `TabView` is a `LinearLayout` we actually hit a limitation in our `SnapshotProducer` architecture: when a `ViewGroup` implementation is mapped entirely through a `WireframeMapper` we need to stop traversing it and just return. To address this problem we had to modify the 
architecture and introduce the `ViewTreeTraversal` that will coordinate the way a `ViewGroup` is traversed and mapped.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: library/dd-sdk-android-session-replay-material/src/test/kotlin/com/datadog/android/sessionreplay/material/forge/TextWireframeForgeryFactory.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.material.forge
```
Comment (0xnm) [inline library/dd-sdk-android-session-replay-material/src/test/kotlin/com/datadog/android/sessionreplay/material/forge/TextWireframeForgeryFactory.kt:7]: should it be in the `material` package? seems like nothing from Material in this class.

File: library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/TreeViewTraversal.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    private val viewUtilsInternal: ViewUtilsInternal = ViewUtilsInternal()
+) {
+
+    @Suppress("ReturnCount")
+    fun traverse(
+        view: View,
+        systemInformation: SystemInformation
+    ): TraversedTreeView {
+        if (viewUtilsInternal.checkIfNotVisible(view) ||
+            viewUtilsInternal.checkIfSystemNoise(view)
```
Comment (0xnm) [inline library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/TreeViewTraversal.kt:30]: should we simply rename methods to `isNotVisible` and `isSystemNoise`?

File: library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/TreeViewTraversal.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        ) {
+            return TraversedTreeView(emptyList(), TraversalStrategy.STOP_AND_DROP_NODE)
+        }
+
+        if (viewUtilsInternal.checkIsToolbar(view)) {
```
Comment (0xnm) [inline library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/TreeViewTraversal.kt:35]: same here. `isToolbar`?

**Review by xgouchet — APPROVED** [review]

File: library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/BaseWireframeMapper.kt [hunk 4]
```diff
@@ -17,21 +17,33 @@ import com.datadog.android.sessionreplay.model.MobileSegment
...
+     * Resolves the [View] unique id to be used in the mapped [MobileSegment.Wireframe].
+     */
     protected fun resolveViewId(view: View): Long {
         // we will use the System.identityHashcode in here which always returns the default
         // hashcode value whether or not a child class overrides this.
         return System.identityHashCode(view).toLong()
     }
 
+    /**
+     * Takes color and the alpha value and returns a html formatted color.
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/mapper/BaseWireframeMapper.kt:36]: Good point, I will try to add something better there. the format might be a better idea actually.

File: library/dd-sdk-android-session-replay-material/src/test/kotlin/com/datadog/android/sessionreplay/material/forge/TextWireframeForgeryFactory.kt (hunk 1 above)
Comment (mariusc83) [reply library/dd-sdk-android-session-replay-material/src/test/kotlin/com/datadog/android/sessionreplay/material/forge/TextWireframeForgeryFactory.kt:7]: the `material` packages is the main package of this new `module`.

File: library/dd-sdk-android-session-replay-material/src/main/kotlin/com/datadog/android/sessionreplay/material/TabWireframeMapper.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        val selectorId = uniqueIdentifierGenerator.resolveChildUniqueIdentifier(
+            view,
+            SELECTED_TAB_INDICATOR_KEY_NAME
+        ) ?: return null
+        val screenDensity = systemInformation.screenDensity
+        val viewBounds = viewUtils.resolveViewGlobalBounds(view, screenDensity)
+        val selectionIndicatorHeight = SELECTED_TAB_INDICATOR_HEIGHT_IN_PX
+            .densityNormalized(screenDensity)
+        val paddingLeft = view.paddingStart.toLong().densityNormalized(screenDensity)
+        val paddingRight = view.paddingEnd.toLong().densityNormalized(screenDensity)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply library/dd-sdk-android-session-replay-material/src/main/kotlin/com/datadog/android/sessionreplay/material/TabWireframeMapper.kt:59]: But I am going to rename those as `paddingStart` and `paddingEnd`

**Pull request merged**